tests/_client.lua

```diff
@@ -3,27 +3,103 @@ local spawn = require("awful.spawn")
 -- This file provide a simple, yet flexible, test client.
 -- It is used to test the `awful.rules`
 
-return function(class, title, use_sn)
-    class = class or 'test_app'
-    title = title or 'Awesome test client'
-
-    local cmd = {"lua" , "-e", table.concat {
-        "local lgi = require 'lgi';",
-        "local Gtk = lgi.require('Gtk');",
-        "Gtk.init();",
-        "local class = '",class,"';",
-        "local window = Gtk.Window {",
-        "    default_width  = 100,",
-        "    default_height = 100,",
-        "    on_destroy     = Gtk.main_quit,",
-        "    title          = '",title,
-        "'};",
-        "window:set_wmclass(class, class);",
-        "window:show_all();",
-        "Gtk:main{...}"
-    }}
-
-    return spawn(cmd, use_sn)
+local test_client_source = [[
+local lgi = require 'lgi'
+local Gtk = lgi.require('Gtk')
+local Gio = lgi.require('Gio')
+Gtk.init()
+
+local function open_window(class, title, snid)
+    local window = Gtk.Window {
+        default_width  = 100,
+        default_height = 100,
+        title          = title
+    }
+    if snid ~= "" then
+        window:set_startup_id(snid)
+    end
+    window:set_wmclass(class, class)
+    window:show_all()
+end
+
+-- Start a coroutine for nicer input handling
+local coro = coroutine.wrap(function()
+    while true do
+        local class = coroutine.yield()
+        local title = coroutine.yield()
+        local snid = coroutine.yield()
+        open_window(class, title, snid)
+    end
+end)
+coro()
+
+-- Read lines from stdin and feed them to the coroutine
+local stdin = Gio.UnixInputStream.new(0, false)
+stdin = Gio.DataInputStream.new(stdin)
+
+local read_start, read_finish
+read_start = function()
+    stdin:read_line_async(0, nil, read_finish)
+end
+read_finish = function(...)
+    local line, length = stdin.read_line_finish(...)
+    if type(length) ~= "number" then
+        error("Error reading line: " .. tostring(length))
+    end
+
+    local eof = not line -- Behaviour of up-to-date LGI
+            or (tostring(line) == "" and #line ~= length) -- (Ab)use uninitialized variable
+    if eof then
+        Gtk.main_quit()
+    else
+        coro(tostring(line)) -- tostring() needed for older LGI versions
+        read_start()
+    end
+end
+
+read_start()
+Gtk:main{...}
+
+-- vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
```

Comment:
Nope. :-)
I wrote that file by hand at first and out of a habbit added a vim modeline. Then I copied the whole file into this Lua file. It could be useful if anyone copies out this chunk of code in its own file to play with it and edit it...
Do you want me to remove this modeline?